Deprecate use of CHANGELOG.md

chosak opened the pull request:
Because someone's gotta do it.

This PR deprecates CHANGELOG.md as the place for tracking changes from release to release in favor of pull request titles.

## Removals

- CHANGELOG.md.

## Changes

- Pull request template now reminds people to give their PRs informative and human-readable titles.

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
* [ ] Reviewers requested with the [Assignee tool](https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/) :arrow_right:

richaagarwal commented:
:ship: it

sebworks commented:
👍  💯  :1st_place_medal: 🔢

rosskarchner commented:
Ohhhh, oh, oh, oh, ohhh.
It must have been cold there in my shadow,
To never have sunlight on your face.
You were content to let me shine, that's your way.
You always walked a step behind.

So I was the one with all the glory,
While you were the one with all the strength.
A beautiful face without a name for so long.
A beautiful smile to hide the pain.

Did you ever know that you're my hero,
And everything I would like to be?
I can fly higher than an eagle,
For you are the wind beneath my wings.

It might have appeared to go unnoticed,
But I've got it all here in my heart.
I want you to know I know the truth, of course I know it.
I would be nothing without you.

Did you ever know that you're my hero?
You're everything I wish I could be.
I could fly higher than an eagle,
For you are the wind beneath my wings.

Did I ever tell you you're my hero?
You're everything, everything I wish I could be.
Oh, and I, I could fly higher than an eagle,
For you are the wind beneath my wings,
'cause you are the wind beneath my wings.

Oh, the wind beneath my wings.
You, you, you, you are the wind beneath my wings.
Fly, fly, fly away. You let me fly so high.
Oh, you, you, you, the wind beneath my wings.
Oh, you, you, you, the wind beneath my wings.

Fly, fly, fly high against the sky,
So high I almost touch the sky.
Thank you, thank you,
Thank God for you, the wind beneath my wings.

Scotchester commented:
:-1:

Pull request titles don't tell you when something was released, so this change will make it harder to figure out what is/isn't released (or when something was released) until we get to a more CD state.

chosak commented:
@Scotchester would using the script [in this PR](https://github.com/cfpb/cfgov-refresh/pull/2723) satisfy those concerns, as long as release notes contain the list of changes contained within them? I believe @willbarton created the last few [release notes](https://github.com/cfpb/cfgov-refresh/releases) this way.

Scotchester commented:
@chosak Yes, if that script is run, has useful PR titles to draw from, and is consistently posted to the GitHub release notes, that's good enough. For extra credit, you could put the list in the Git tag's [annotations](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags), which would ensure we're not dependent on GitHub to have access to that info in the future. I'm fairly sure that a tag annotation will show up in the GitHub release description when the tag is pushed.

I might also suggest that the existing CHANGELOG file be retained in the repo (possibly renamed) with a note that all future changes are noted in the tag/release.

willbarton commented:
> I might also suggest that the existing CHANGELOG file be retained in the repo (possibly renamed) with a note that all future changes are noted in the tag/release.
 
It's in the git history.

Murder it.

chosak commented:
@Scotchester I like the annotated tag idea. How would you and @willbarton feel about archiving a copy of the CHANGELOG in the docs folder? Leaving it around in the main repo root directory seems prone to confusion.

Scotchester commented:
@chosak I'd prefer that over straight deletion, but I think simply renaming to something like `CHANGELOG_archive.md` would be more obvious to people who default to looking for the standard `CHANGELOG.md`.

willbarton commented:
@Scotchester How would that help though? I'd be open to having a `CHANGELOG.md` that simply explained where to find the latest changes, and perhaps included the URL of the last commit before this one for historical changes. Wouldn't that suit someone who defaults to looking for a file named something like `CHANGELOG` better? Why do we need to preserve the full contents for posterity in the tree when it will be in the git history?

Scotchester commented:
OK, if you include a link to the last commit that had the change log content, so someone can easily get back to it, I'm cool with that. Leaving it named `CHANGELOG.md` is even better.

My main concern was with deleting the file entirely with no guidance on how to get back to that information. Then, to find it in the history, one has to (a) know that the file used to exist, and (b) have to dig back through however many commits have occurred since to find the last version of it.

I fully support your latest proposal, @willbarton.

willbarton commented:
![5d93c04122701e89f2d839d0e80e5a71](https://cloud.githubusercontent.com/assets/10562538/22649400/b774a6e2-ec48-11e6-9991-17e9668c6e19.gif)

chosak commented:
@Scotchester how do you feel about [something like this](https://github.com/cfpb/cfgov-refresh/blob/die-changelog-die/CHANGELOG.md) as the new placeholder `CHANGELOG.md`? Note that I've had to hardcode a version there so we'll have to keep updating that if we keep this PR open for a long time.

---

Also, at your suggestion I looked into annotated tags. It seems like we can choose to use either GitHub releases or git annotated tags.

If we use GitHub releases, then (as now), you won't be able to see the [full release description](https://github.com/cfpb/cfgov-refresh/releases/tag/4.6.7) from the command line, only the first few lines:

```sh
$ git tag -n99 -l 4.6.7
4.6.7           Merge pull request #2724 from cfpb/build-fix
    
    Fix npm dependencies and shrinkwrap for production deploys
```

If we use annotated git tags, then from the command line we can do something like

```sh
$ git tag -a 100.0.2 # opens editor where multi-line annotation can be entered
$ git push origin 100.0.2
```

The full content can then be viewed on the repo [tags page](https://github.com/cfpb/cfgov-refresh/tags):

![image](https://cloud.githubusercontent.com/assets/654645/22669273/ddbe79a6-ec91-11e6-8b44-da0ce51c0743.png)

Note the "Add release notes" prompt. If you enter release notes, then they get added as additional content and can be viewed on the [releases page](https://github.com/cfpb/cfgov-refresh/tags), while the tag content can still be viewed on the tags page.

Based on this, I'd be open to using annotated git tags over GitHub releases going forward. The ability to view tag descriptions from command line is nice, and it'd likely be easier to integrate with an automated tag-and-deploy process (no need to use the GH API).

Scotchester commented:
Looking good, Andy. I'll make a few trivial style suggestions on the `CHANGELOG.md` file in a sec.

On annotations: If you click that "Add release notes" link, does it do what I hope and prepopulate the release notes field with the content of the tag annotation?

Scotchester commented on CHANGELOG.md:
```diff
@@ -1,2128 +1,13 @@
-All notable changes to this project will be documented in this file.
+This project no longer uses CHANGELOG.md to track release changes.
```
Trivial: Wrap `CHANGELOG.md` in backticks.

Scotchester commented on CHANGELOG.md:
```diff
@@ -1,2128 +1,13 @@
-- Transitioned Capital Framework dependency to v1.0.0 in bower.json.
-- Added gulp and the required npm plugins
-- Added gulp config file to lay out configs for each task
-- Added gulp task
```
Wrap this line in backticks or a shell code fence.

Scotchester replied:
Sorry, put this on the wrong line. It should be referring to the shell command line.

chosak commented:
@Scotchester thanks for the feedback.

Regarding release notes, unfortunately, no, these do not prepopulate from tag annotations.

I've addressed your comments in 6515cb6. If you think these look good, I'll merge this PR.

Scotchester commented:
:shipit:

Also, if we're going to be building the notes from PR titles, you'll want to update the title of this PR to reflect what it actually changes now :)
